pkg/test/server.go

```diff
@@ -0,0 +1,156 @@
+package test
+
+import (
+	"encoding/csv"
+	"fmt"
+	"log"
+	"net"
+	"os"
+
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+var domainToIp = loadRecords("data/test-300.csv")
+
+const truncateSize = 512
+
+type Server struct {
+	Addr    string
+	TcpPort int
+	UdpPort int
+}
+
+func (s *Server) parseUDPAddr() string {
+	return fmt.Sprintf("%v:%v", s.Addr, s.UdpPort)
+}
+
+func (s *Server) parseTCPAddr() string {
+	return fmt.Sprintf("%v:%v", s.Addr, s.TcpPort)
+}
+
+func (s *Server) Run() {
+	g := new(errgroup.Group)
+	udpSrv := &dns.Server{Addr: s.parseUDPAddr(), Net: "udp", Handler: &udpHandler{}}
+	defer func() {
+		err := udpSrv.Shutdown()
+		if err != nil {
+			logrus.Errorf("shutting down UDP server: %v", err)
+		}
+	}()
+	g.Go(func() error {
+		return udpSrv.ListenAndServe()
+	})
+	tcpSrv := &dns.Server{Addr: s.parseTCPAddr(), Net: "tcp", Handler: &tcpHandler{}}
+	defer func() {
+		err := tcpSrv.Shutdown()
+		if err != nil {
+			logrus.Errorf("shutting down TCP server: %v", err)
+		}
+	}()
+	g.Go(func() error {
+		return tcpSrv.ListenAndServe()
+	})
+
+	if err := g.Wait(); err != nil {
+		logrus.Fatalf("Failed starting servers: %v", err)
+	}
+}
+
+type tcpHandler struct{}
+
+func (h *tcpHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	msg := new(dns.Msg)
+	msg.SetReply(r)
+	msg.Compress = false
+	// we might handle other OpCodes later
+	if r.Opcode != dns.OpcodeQuery {
+		return
+	}
+	parseReply(msg)
+	err := w.WriteMsg(msg)
+	if err != nil {
+		logrus.Errorf("ServeDNS trying to write message in tcpHandler: %v", err)
+	}
+}
+
+type udpHandler struct{}
+
+func (h *udpHandler) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	msg := new(dns.Msg)
+	msg.SetReply(r)
+	msg.Compress = false
+	// we might handle other OpCodes later
+	if r.Opcode != dns.OpcodeQuery {
+		return
+	}
+	parseReply(msg)
+	msg.Truncate(truncateSize)
+	err := w.WriteMsg(msg)
+	if err != nil {
+		logrus.Errorf("ServeDNS trying to write message in udpHandler: %v", err)
+	}
+}
+
+func parseReply(msg *dns.Msg) {
+	logrus.Infof("ServeDNS handling request: %v", msg.Question)
+	for _, q := range msg.Question {
+		switch q.Qtype { //nolint:gocritic
+		case dns.TypeA:
+			msg.Authoritative = true
+			domain := q.Name
+			addresses, ok := domainToIp[domain]
+			if ok {
+				for _, addr := range addresses {
+					msg.Answer = append(msg.Answer, &dns.A{
+						Hdr: dns.RR_Header{
+							Name:   domain,
+							Rrtype: dns.TypeA,
+							Class:  dns.ClassINET,
+							Ttl:    60,
+						},
+						A: net.ParseIP(addr),
+					})
+				}
+			}
+		}
+	}
+}
+
+func loadRecords(path string) map[string][]string {
```

Comment:
Not too sure at this point which path we will take, either static data or generating data on the fly. For now, we only have A records but that will change shortly as we need to add other RRs. Maybe, we can make this decision then since it is a simple change.